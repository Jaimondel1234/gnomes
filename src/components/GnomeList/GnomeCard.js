import React from 'react';
import {
  PanelStyle,
  ImageStyle,
  FlexStyle,
  ContainerStyle,
  ButtonStyle,
  SpanStyle,
} from '../../styles/core';
import { useHistory } from 'react-router-dom';
import PropTypes from 'prop-types';

/*
  Component that displays the gnome card
*/
export const GnomeCard = ({ ...props }) => {
  const { name, thumbnail } = props;
  const history = useHistory();
  const handleClick = () => {
    history.push('/gnome-detail', { ...props });
  };

  return (
    <PanelStyle
      mr='30px'
      mb='50px'
      maxWidth='900px'
      borderRadius='10px'
      pt='20px'
      mt='10px'
    >
      <FlexStyle
        alignItems='center'
        justifyContent='space-between'
        flexDirection='column'
      >
        <SpanStyle
          color='white'
          fontSize='30px'
          fontWeight='bold'
          p='15px'
          textShadow='2px 2px 4px #000000'
          borderRadius='5px'
          bg='#1E90FF'
        >
          {name}
        </SpanStyle>
        <ContainerStyle p='20px'>
          <ImageStyle
            src={thumbnail}
            width='350px'
            height='300px'
            onError={(e) => {
              e.target.onerror = null;
              e.target.src =
                'http://www.publicdomainpictures.net/pictures/30000/nahled/maple-leaves-background.jpg';
            }}
          />
        </ContainerStyle>
        <ButtonStyle
          primary
          p='12px'
          fontSize='25px'
          fontWeight='bold'
          onClick={handleClick}
          letterSpacing='3px'
          lineHeight='50px'
        >
          ... More
        </ButtonStyle>
      </FlexStyle>
    </PanelStyle>
  );
};

GnomeCard.propTypes = {
  props: PropTypes.shape({
    name: PropTypes.string,
    thumbnail: PropTypes.string,
    age: PropTypes.number,
    genre: PropTypes.string,
    hair_color: PropTypes.string,
    height: PropTypes.number,
    weight: PropTypes.number,
    professions: PropTypes.arrayOf(PropTypes.string),
    friends: PropTypes.arrayOf(PropTypes.string),
  }),
};
