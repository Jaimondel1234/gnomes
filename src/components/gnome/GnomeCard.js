import React from "react";
import {
  PanelStyle,
  TextStyle,
  ImageStyle,
  FlexStyle,
  ContainerStyle,
  ButtonStyle,
} from "../../styles/core";

export const GnomeCard = ({ name, thumbnail }) => {
  return (
    <PanelStyle
      bg="#1E90FF"
      mr="30px"
      mb="50px"
      maxWidth="900px"
      borderRadius={10}
      pt="50px"
    >
      <FlexStyle
        alignItems="center"
        justifyContent="space-between"
        flexDirection="column"
      >
        <TextStyle
          color="white"
          fontSize="30px"
          fontWeight="bold"
          textShadow="2px 2px 4px #000000"
        >
          {name}
        </TextStyle>
        <ContainerStyle p="20px">
          <ImageStyle
            src={thumbnail}
            width="350px"
            height="300px"
            onError={(e) => {
              e.target.onerror = null;
              e.target.src =
                "http://www.publicdomainpictures.net/pictures/30000/nahled/maple-leaves-background.jpg";
            }}
          />
        </ContainerStyle>
        <ButtonStyle primary p="12px" fontSize="22px" fontWeight="bold">
          Send
        </ButtonStyle>
      </FlexStyle>
    </PanelStyle>
  );
};
