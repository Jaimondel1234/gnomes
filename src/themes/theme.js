/* 
  Theme
*/

export const theme = {
  color: {
    main: '#007bff',
  },
};
