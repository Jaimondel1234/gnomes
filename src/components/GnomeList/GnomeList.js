import React, { useRef, useState, useEffect } from "react";
import {
  FlexStyle,
  ScrollContainer,
  PlaceHolderStyle,
} from "../../styles/core";
import { GnomeCard } from "./GnomeCard";
import { useSelector, useDispatch } from "react-redux";
import { loadGnomes } from "../../actions/gnome";
import { fill } from "lodash";
import map from "lodash/map";

let emptyArray = fill(new Array(20), 0);

const placeHolderArray = emptyArray.map((value, id) => {
  return `place_holder_${id}`;
});

export const GnomeList = () => {
  const dispatch = useDispatch();
  const { gnomes, loading } = useSelector((state) => state.gnome);
  const ref = useRef(null);

  const handleScroll = () => {
    if (ref.current.scrollTop >= 0.75 * ref.current.scrollHeight) {
      dispatch(loadGnomes());
    }
  };
  return (
    <ScrollContainer ref={ref} onScroll={handleScroll}>
      <FlexStyle
        alignItem="center"
        justifyContent="center"
        flexDirection="row"
        flexWrap="wrap"
      >
        {gnomes &&
          map(gnomes, ({ id, ...props }) => <GnomeCard key={id} {...props} />)}
        {loading &&
          map(placeHolderArray, (value) => (
            <PlaceHolderStyle
              key={value}
              bg="grey"
              mr="30px"
              mb="50px"
              maxWidth="900px"
              borderRadius={10}
              pt="50px"
            />
          ))}
      </FlexStyle>
    </ScrollContainer>
  );
};
