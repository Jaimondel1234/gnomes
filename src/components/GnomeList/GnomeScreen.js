import React, { useEffect } from "react";
import { ContainerStyle, ScrollContainer } from "../../styles/core";
import { GnomeList } from "./GnomeList";
import { loadGnomes } from "../../actions/gnome";
import { useDispatch, useSelector } from "react-redux";

export const GnomeScreen = () => {
  const dispatch = useDispatch();
  const { scrollNumberGnomes } = useSelector((state) => state.gnome);

  useEffect(() => {
    if (scrollNumberGnomes === 0) {
      dispatch(loadGnomes());
    }
  }, [loadGnomes, dispatch, scrollNumberGnomes]);

  return (
    <ContainerStyle mt="5">
      <GnomeList />
    </ContainerStyle>
  );
};
