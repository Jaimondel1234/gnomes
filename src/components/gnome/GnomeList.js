import React, { useRef, useEffect } from "react";
import { FlexStyle, ScrollContainer } from "../../styles/core";
import { GnomeCard } from "./GnomeCard";
import { useSelector, useDispatch } from "react-redux";
import { loadGnomes } from "../../actions/gnome";

export const GnomeList = () => {
  const dispatch = useDispatch();
  const { gnomes } = useSelector((state) => state.gnome);
  const ref = useRef(null);

  const handleScroll = () => {
    if (ref.current.scrollTop >= 0.75 * ref.current.scrollHeight) {
      dispatch(loadGnomes());
    }
  };
  return (
    <>
      <ScrollContainer ref={ref} onScroll={handleScroll}>
        <FlexStyle
          alignItem="center"
          justifyContent="center"
          flexDirection="row"
          flexWrap="wrap"
        >
          {gnomes &&
            gnomes.map(({ id, ...props }) => <GnomeCard key={id} {...props} />)}
        </FlexStyle>
      </ScrollContainer>
    </>
  );
};
