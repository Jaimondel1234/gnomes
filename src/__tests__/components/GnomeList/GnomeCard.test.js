import React from 'react';
import { mount } from 'enzyme';
import { ThemeProvider } from 'styled-components';
import { theme } from '../../../themes/theme';
import { GnomeCard } from '../../../components/GnomeList/GnomeCard';
import { MemoryRouter, Router } from 'react-router-dom';
import { gnomesMock } from '../../../__mocks__/gnomesMock';

describe('Test <GnomeCard/>', () => {
  const history = {
    push: jest.fn(),
    location: { state: {} },
    listen: jest.fn(),
    createHref: jest.fn(),
  };
  const props = gnomesMock[0];
  const wrapper = mount(
    <ThemeProvider theme={theme}>
      <MemoryRouter>
        <Router history={history}>
          <GnomeCard props={props} />
        </Router>
      </MemoryRouter>
    </ThemeProvider>,
  );
  test('it must be displayed correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  test('if you click on the button more you must make a redirection with the information of the gnome', () => {
    wrapper.find('button').prop('onClick')();
    expect(history.push).toHaveBeenCalledWith('/gnome-detail', { props });
  });
});
