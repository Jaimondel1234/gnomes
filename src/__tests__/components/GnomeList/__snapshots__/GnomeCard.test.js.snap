// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <GnomeCard/> it must be displayed correctly 1`] = `
<div
  className="core__PanelStyle-qu5lcc-1 hCLCiU"
>
  <div
    className="core__FlexStyle-qu5lcc-4 joXSZK"
  >
    <span
      className="core__SpanStyle-qu5lcc-9 koWOQR"
      color="white"
      fontSize="30px"
      fontWeight="bold"
    />
    <div
      className="core__ContainerStyle-qu5lcc-0 cjWygG"
    >
      <img
        className="core__ImageStyle-qu5lcc-7 ecLPaO"
        height="300px"
        onError={[Function]}
        width="350px"
      />
    </div>
    <button
      className="core__ButtonStyle-qu5lcc-6 eFvMQp"
      fontSize="25px"
      fontWeight="bold"
      letterSpacing="3px"
      onClick={[Function]}
    >
      ... More
    </button>
  </div>
</div>
`;
